### PR TITLE
Add retry logic

### PR DESCRIPTION
**Pull request by kittylyst**

### CODE REVIEW

File: jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java [hunk 1]
```diff
@@ -20,6 +21,7 @@
   private static final Logger logger = LoggerFactory.getLogger(JFRJMXRecorder.class);
 
   private static final int MAX_BYTES_READ = 5 * 1024 * 1024;
+  private static List<Integer> backoffSeconds = List.of(1, 2, 4, 8, 15);
```
Comment (breedx-nr) [inline jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java:24]: should be final, and convention would have it be all caps.

File: jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java [hunk 2]
```diff
@@ -34,14 +36,55 @@ public JFRJMXRecorder(
...
+          // Log retry?
+          if (i == backoffSeconds.size() - 1) {
+            throw iox;
+          }
+        }
+      }
+    }
+
+    return out;
+  }
```
Comment (breedx-nr) [inline jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java:58]: The single-point return spot and the nested try/catch make this hard to follow....but most importantly, I think it's also missing a sleep!  What do you think about something like this?

```suggestion
  public static JFRJMXRecorder connectWithBackOff(DaemonConfig config) throws IOException {
    return connectWithBackOff(config, 0);
  }

  private static JFRJMXRecorder connectWithBackOff(DaemonConfig config, int backoffIndex) throws IOException {
    try {
      return connect(config);
    } catch (IOException e) {
      if(backoffIndex == backoffSeconds.size() - 1) {
        logger.error("Failed to connect to JMX and retries exhausted.  JFR will be disabled.", e);
        throw e;
      }
      logger.warn("Error connecting to JMX, backing off before retry", e);
      sleepSafely(backoffIndex);
      return connectWithBackOff(config, backoffIndex + 1);
    }
  }

  private static void sleepSafely(int backoffIndex) {
    try {
      TimeUnit.SECONDS.sleep(backoffSeconds.get(backoffIndex));
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new RuntimeException("Error during backoff sleep", e);
    }
  }
```

File: jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java [hunk 3]
```diff
@@ -34,14 +36,55 @@ public JFRJMXRecorder(
...
+
+  public void startRecordingWithBackOff()
+      throws MalformedObjectNameException, ReflectionException, IOException,
+          InstanceNotFoundException, MBeanException, OpenDataException {
+    try {
+      startRecording();
+    } catch (Exception e) {
+      for (int i = 0; i < backoffSeconds.size(); i = i + 1) {
```
Comment (breedx-nr) [inline jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java:66]: I would prefer not to see the same construct repeated.  The recursive approach might help, but maybe it warrants an abstraction?

Comment (kittylyst) [reply jfr-daemon/src/main/java/com/newrelic/jfr/daemon/JFRJMXRecorder.java]: I, too, would prefer that this construct not be repeated. However, I had a look at it and I suspect that abstracting it is trickier than it appears (e.g. due to differences in exception signature) and the resulting code is going to look damned ugly and hard to read. This is the kind of situation where I wish we had more of a functional language to work in - if this was Scala (or Kotlin?) then this would become a "with" pattern, but we are where we are.

This was quite a long-winded way of saying that I agree, but in the interests of getting this feature out to users, I've gone with the repeated construct for now. I'm up for a tidied-up functionalist version if it's not too complex or ugly, but let's get this in sooner.
